### ISSUE TITLE
Need mechanism for libs to use code from "ancestor" libs

### ISSUE THREAD

**Issue by aerwin** [maqetta/maqetta#2067]
@jhpedemonte 

This is a spin-off from #2034. The summary is from the new gridx plug-in I want to use some modules from the Dojo 1.7 plugin. I was able to make this work running from Eclipse with a relative path. But, then when I run from a build, that relative path fails.

I can keep making gridx progress from Eclipse, but we'll need to resolve this before it will work from builds. So, making priority high.

A chat transcript between myself and @jhpedemonte follows:

```
Anthony : Hey, Javier... Good morning... wondered if I could ask a question related to referencing classes in another plugin? 
Javier : sure 
Javier : i'll try my best to answer 
Javier : :-) 
Anthony : thanks  :-) From a class in my new gridx plugin I want to make use of DataGridHelper in the Dojo 1.7 plugin, and the following works when run from Eclipse: 
Anthony : define([
    "../../metadata/dojo/1.7/dojox/grid/DataGridHelper"
], function(
    DataGridHelper
) { 
Anthony : BUT, when I run from a build, that relative path fails 
Javier : hmm, interesting 
Anthony : do I have any alternatives? 
Javier : i've never considered this possibility 
Javier : maybe 
Javier : i'll have to take a look, dig into the code 
Anthony : yeah... didn't see anyone else doing it 
Javier : is this out on a branch? 
Anthony : I can push it out 
Javier : yes, please 
Javier : and i'll dig in and see what i can find 
Anthony : awesome... thanks 
Anthony : ok... out in the 2034_initial_gridx_work branch 
Anthony : you went offline right as I sent this earlier, so wanted to make sure you got it: "ok... out in the 2034_initial_gridx_work branch" 
Javier : yeah, i saw it 
Anthony : ok. cool 
Anthony : thanks 
Anthony : no rush 
Javier : i think we need to come up with a mechanism or API to allow dependent libraries to make use of the resources in their "ancestor" libs. 
Javier : but i'll have to think on that 
Javier : do you have a bug open for this? 
Anthony : no 
Anthony : but I can open one 
```

**Comment by jhpedemonte** [comment]
(1) In **metadata.js**, in order to load helpers, create tools, etc, we create a **module ID** (namespace) for each library's metadata objects.  This makes it easier to load using AMD.  In the case of the Dojo library, it's module ID is **"maq-metadata-dojo-1.7"**.

So, one easy solution would be to use this module ID in dependent libraries:

``` javascript
define([
    "maq-metadata-dojo-1.7/dojox/grid/DataGridHelper"
], function(DataGridHelper) { 
```

(2) The issue with that approach is that the module ID **contains the version number** of the library.  This makes it less than ideal, since the "base" library can be updated independently from the other library.  We need an approach that doesn't presume a particular version.

We could make use of `metadata.getHelper()`, like so:

``` javascript
metatadata.getHelper('dojox.grid.DataGrid', 'tool').then(function(module) {
    // 'module' == DataGridCreateTool obj
    // can now make use of it in this func
});
```

While this would work in a lot of cases, it won't work for this use, since we are trying to create (and immediately return as AMD module) an object that inherits from `DataGridCreateTool`.  That means that we can't use any **async** calls.

(I played around with the idea of creating a façade that would allow the GridX create tool obj to lazy load DataGridCreateTool and then call the appropriate inherited functions at the right time.  But it doesn't work (well) for the constructor. Plus, this isn't a good strategy going forward.  :-)

---

So the best solution then is to create a extensible solution for the **module ID**.

I originally added the version number into the ID under the assumption that we would (at some point) allow multiple versions of libraries in Maqetta, and that those different versions could be loaded at once (i.e. different files open at same time using different versions of same lib -- something we had previously discussed).  If this is something we no longer want, then we can remove the version.

Another concern is how we handle "split" libraries -- where the code and metadata can be stored in separate packages (see the [Library Packages](https://github.com/maqetta/maqetta/wiki/Library-Packages) wiki for details).  The "dojo" library could actually consist of 3 different packages: "dojo" (core toolkit code), "dojo-oam" (OpenAjax metadata for widgets), "dojo-maq" (Maqetta-specific metadata).

The GridX library shouldn't have to worry about whether it needs to depend on an all-in-one "dojo" or "dojo-maq".  Given that, is there a way to construct the **module ID** (i.e. "maq-metadata-dojo", or something along those lines) that includes all of the relevant metadata?

@peller @JonFerraiolo

**Comment by JonFerraiolo** [comment]
Sorry I'm slow in responding. Maybe this has been resolved, but in case it hasn't, here is another idea. 

You know how sometime people set up URLs with symbolic links such that particular versions are stored at `<basedir>/v1.1/` and `<basedir>/v1.2/` and maybe there is the latest version at something like `<basedir>/latest/` which under the hood is a symlink or redirect to the latest version? I wonder if we can use a similar approach for module IDs, where we have module IDs for each version and another module ID that represents the latest version. 

It seems to me that a significant percentage of the time, developers will want to point to a particular version versus latest version because they have only verified their code against a particular version, whereas other times developers will say the want to use the latest. I don't think we can cover all scenarios if we just pick one approach (eg. always supply version) or another (latest version).

One way to achieve this would be to define a trivial library package that has nothing in it except for the dependency definition on the latest version of that library package. For example, davinci.dojo_latest => davinci.dojo_1.7.

**Comment by JonFerraiolo** [comment]
-> @jhpedemonte (so this doesn't fall through cracks)

**Comment by jhpedemonte** [comment]
Nope, not fixed yet.

**Comment by jhpedemonte** [comment]
Now closing. Fixed by 581272d90df90bb6c725481de48ced56117b24b3.

Only issue I see is an error trying to get `http://localhost:8081/maqetta/app/static/lib/gridx/dojox/html/resources/ellipsis.css`.  But that should be a separate issue.

**Comment by aerwin** [comment]
If you pull the latest GridX submodule, the ellipsis issue should be fixed (#2172).

**Comment by jhpedemonte** [comment]
A submodule is actually a link to a specific revision/commit.  So if you update the submodule, you'll have to commit it, as you would a normal file.  Then, when the rest of us pull, it should make us pull the updated GridX code as well.
